### ISSUE TITLE
Any "skip_list" causes ansible-lint to give a return code of 0

### ISSUE THREAD

**Issue by jalanb** [ansible-community/ansible-lint#1102]
##### Summary

Ansible-lint does not fail (have a non-zero return code) despite "violation(s) that are fatal" when the file `.ansible-lint` exists and contains `skip_list`. Even when the error codes listed in the `skip_list` are irrelevant.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
ansible --version
ansible 2.10.3
  config file = /opt/wwts/clones/gitlab/bots/deploys/develop/ansible.cfg
  configured module search path = ['/Users/jab/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/wwts/clones/gitlab/bots/deploys/develop/.venv/lib/python3.8/site-packages/ansible
  executable location = /opt/wwts/clones/gitlab/bots/deploys/develop/.venv/bin/ansible
  python version = 3.8.5 (v3.8.5:580fbb018f, Jul 20 2020, 12:11:27) [Clang 6.0 (clang-600.0.57)]

ansible-lint --version
ansible-lint 4.3.8.dev3+g6588df2
```

- ansible installation method: one of source, pip, OS package
pip
- ansible-lint installation method: one of source, pip, OS package
source (current master @ `6588df23735f0c5a51630d7cff6bed6f4b2a48c6`)

##### OS / ENVIRONMENT

macOS Catalina 10.15.7

##### STEPS TO REPRODUCE

Create a new empty directory, run `$ git init`
Add file `provision.yaml` wth contents
```yaml
---

- hosts: localhost
  connection: local

  tasks:
    - name: Copy this
      template:
        src: provision.yaml
        dest: /etc/provision.yaml
```
Add file to git (`$ git add .`)

Run ansible-lint:
```console
$ ansible-lint -pq provision.yaml && echo PASS || echo FAIL
WARNING  Listing 1 violation(s) that are fatal
[208] provision.yaml:7
FAIL
```

That "FAIL" is expected - `208` is "File permissions unset or incorrect", which is the case

Now add the file `.ansible-lint` with contents, which are **not relevant to the errors** seen (only incude code `403`, **not** `208`). 

```yaml
--- 

skip_list:
  - '403'
```
And lint again
```console
$ ansible-lint -pq provision.yaml && echo PASS || echo FAIL
WARNING  Listing 1 violation(s) that are fatal
[208] provision.yaml:7
PASS
```
##### Desired Behaviour

Expected the second run to also show `FAIL`, same as first run

##### Actual Behaviour

It showed `PASS`

I also tried with unrecognised ids instead of `403`. 
_Any_ values in the skip list, even non-numbers such as `'abc def'`, caused the same behaviour.

**Comment by jalanb** [comment]
Further testing: remove the config file, and use command line arguments instead. That acted as expected showing `FAIL`:
```console
$ rm -f .ansible-lint
removed '.ansible-lint'

$ ansible-lint -x 403 -pq provision.yaml && echo PASS || echo FAIL
WARNING  Listing 1 violation(s) that are fatal
[208] provision.yaml:7
FAIL

$ ansible-lint -x abd -pq provision.yaml && echo PASS || echo FAIL
WARNING  Listing 1 violation(s) that are fatal
[208] provision.yaml:7
FAIL
```

Restoring `.ansible-list`, and then using those `-x` options again: brings back the `PASS`

**Comment by greg-hellings** [comment]
Duplicate of #1081. Here's my setup:

```console
(tools3) [ghelling@work 🎩 network_scripts] (master)$ cat ../../tests/ansible-lint 
skip_list:
  - "204"
(tools3) [ghelling@work 🎩 network_scripts] (master)$ ansible-lint .
WARNING  Listing 1 violation(s) that are fatal
[208] File permissions unset or incorrect
tasks/main.yml:15
Task/Handler: copy found files to temp directory

You can skip specific rules or tags by adding them to your configuration file:                                        

┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│# .ansible-lint                                                                                                     │
│warn_list:  # or 'skip_list' to silence them completely                                                             │
│  - '208'  # File permissions unset or incorrect                                                                    │
│  - experimental  # all rules tagged as experimental                                                                │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(tools3) [ghelling@work 🎩 network_scripts] (master)$ echo $?
2
(tools3) [ghelling@work 🎩 network_scripts] (master)$ ansible-lint -c ../../tests/ansible-lint 
WARNING  Listing 2 violation(s) that are fatal
[208] File permissions unset or incorrect
molecule/shared/prepare.yml:4
Task/Handler: create directory

[208] File permissions unset or incorrect
tasks/main.yml:15
Task/Handler: copy found files to temp directory

You can skip specific rules or tags by adding them to your configuration file:                                        

┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│# .ansible-lint                                                                                                     │
│warn_list:  # or 'skip_list' to silence them completely                                                             │
│  - experimental  # all rules tagged as experimental                                                                │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(tools3) [ghelling@work 🎩 network_scripts] (master)$ echo $?
0
```

**Comment by greg-hellings** [comment]
Interestingly, I'm working on the same error 208 as you're seeing. Adding some extra debugging to the output gives me this interesting tidbit:

```console
(tools3) [ghelling@work 🎩 roles] (master)$ ansible-lint -vvv .
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(colored=True, config_file=None, display_relative_path=True, exclude_paths=[], format='rich', listrules=False, listtags=False, parseable=False, parseable_severity=False, playbook=['.'], progressive=False, quiet=False, rulesdir=[], skip_list=[], tags=[], use_default_rules=False, verbosity=3, warn_list=[])
DEBUG    Loading rules from /home/ghelling/venv/tools3/lib64/python3.7/site-packages/ansiblelint/rules

(tools3) [ghelling@work 🎩 roles] (master)$ ansible-lint -vvv -c ../ansible-lint .
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(colored=True, config_file='../ansible-lint', display_relative_path=True, exclude_paths=[], format='rich', listrules=False, listtags=False, parseable=False, parseable_severity=False, playbook=['.'], progressive=False, quiet=False, rulesdir=[], skip_list=['asb'], tags=[], use_default_rules=False, verbosity=3, warn_list=['experimental'])
DEBUG    Loading rules from /home/ghelling/venv/tools3/lib64/python3.7/site-packages/ansiblelint/rules
```

Note that adding the config file that includes a `skip_list` entry *silently and incorrectly* changes `warn_list` from `[]` to `['experimental']`.

This is wrong and should not happen. But, now that it's happening, I can maybe finally get to the bottom of a test case that demonstrates this. Because most of the test cases are working against items that are not experimental.

**Comment by xenlo** [comment]
Minor detail, but I think there is a typo in the title: `ansible-list` —> `ansible-lint`.

**Comment by jalanb** [comment]
@xenlo 

Thanks, fixed

**Comment by netzvieh** [comment]
@ssbarnea this was introduced in #1037

If there's a config file where `warn_list` is not defined, `warn_list` gets the default of `['experimental']` as desribed in the parameter help. But on the commandline the default is `[]` which leads to different behavior depending on wether we define a `skip_list` in a config file or on the commandline.

What's the expected behavior? Always only warn about `experimental` rules unless the user explicitly disables that - whether through setting `warn_list` in the config file or using `-w` on the commandline?

**Comment by ssbarnea** [comment]
Yep: that was the idea of `experimental`, to act only as warning as long user does not override warn_list. This means that if you want to add stuff to warn_list, you also need to manually add experimental to prevent them from becoming failures.

**Comment by netzvieh** [comment]
And just that I get it correct in the implementation:

If I have list-param X with the following values:

* default: ['a']
* config file: ['b']
* commandline: ['c']

The end-result should be ['c', 'b'], so once either config file or commandline is specified, we drop the default, but we still merge cli and config file?

**Comment by ssbarnea** [comment]
Default is an exception because if we would merge, it would be impossible to remove the implicit (experimental) value from the list. That is why specifying any value in config or command line overrides default. The same logic applies to any default.

Likely the only bug here is that we do not show the implicit value on the command line help.

**Comment by netzvieh** [comment]
Thing is the implicit value is only taken into account once we specify a config file, because if none is specified the function returns before merging the defaults.

No Config? -> Just return CLI values
Yes Config? -> Take CLI values, merge config file values and defaults where nothings in the config file

I'd say we want to also merge the defaults if the commandline arguments are not specified and still have their default values (empty arrays), even if theres no config file.

**Comment by ssbarnea** [comment]
The way to additive commands cli work is that they add values to a list. AFAIK there is no way to pass an empty array/list from the cli. If you find one, I am curious to learn it.
